### DISCUSSION TITLE
[newbies] ways of pulling data from LRGS servers.

### DISCUSSION THREAD

**Discussion by SebastienCharland-HQ** [opendcs/opendcs#790] [Q&A] [answered]
Hi,

We're planning on rewriting our Windows LRGS data pull service based on DDS protocol 14, to something that will run on RedHat.
Hopefully in bash/sftp-rsync with authorized ssh key, or python/libcurl.

Anyone can point me to documents on what is available / possible with the current system?

**Comment by ktarbet** [comment]
If you install/run LRGS , you can configure it to pull data from another LRGS.
This results in a directory with xml files (one per day).

https://opendcs-env.readthedocs.io/en/latest/legacy-lrgs-userguide.html#lrgs-configuration-1

**Reply by SebastienCharland-HQ** [reply]
Excellent link!
I will look into this LRGS local server installation.
I also see that there is this 'getDcpMessages' utility script, which might be more along what I was loking for.
I see that it only uses '-u username', does that mean that the password has to be part of the username string, ie, 'username@password', or do I need a special user configured on the LRGS servers?

**Comment by SebastienCharland-HQ** [comment]
I was able to install and (pseudo)configure an opendcs instance localy, but I couldn't find if there was way to slow down the rate at which my opendcs pool the other LRGS servers? 
I only have a few stations to query every minutes, so pooling every second is too fast.

**Reply by SebastienCharland-HQ** [reply, marked as answer]
I answered myself yesterday; I changeg the timeout value from 1s to 60s in DdsRecv.java allCaughtUp(), recompiled the file and repacked the jar... Works like a charm.

**Reply by MikeNeilson** [reply]
@opendcs/opendcs-core-devs anyone else think perhaps we should just make that configurable? I had to do the same thing to the computation wait loop.

**Comment by MikeNeilson** [comment]
If you wish to make your own client the protocol is defined here: https://github.com/opendcs/dcs_standards/blob/main/source/dds-14.rst

In-progress work will add TLS and STARTTLS support. And we're also creating an HTTP based version that is being defined here: https://github.com/opendcs/dcs_standards/pull/3/files#diff-f37cd5bbd65a5354ab8cf97140790797269da9171917c15c7b2ab15f462e1a00


Also I swear someone created, or was creating a python client and gave us a link to it but for the life of me I can't find it in the discussions.

**Reply by tve** [reply]
Python client: https://github.com/dcspy/dcspy

**Reply by SebastienCharland-HQ** [reply]
You just saved me hours of coding.
If you're ever in Laval/Montréal, coffee is on me!

**Reply by tve** [reply]
Note that I'm not the author of that package! I did submit a PR to fix a bug and it looks like I'll be adding support for network lists as well for the "hanging on" mode, i.e., querying without DRS_UNTIL so the LRGS keeps sending data on the connection as it comes in.
Overall I would describe the package as being beta quality, it needs some more shaking out and additional touches. I added decoding of the DCS/DOMSAT Message header but haven't pushed that into a PR yet.
Let me know how it's working for you and if you find issues, I think we need to pool efforts a bit to get this over the line...

**Reply by SebastienCharland-HQ** [reply]
There's a few places that could benefit from a try-expect encapsulation for better error handling, but overall the package is promising and will definitely give us a leg up.

**Reply by darshanbaral** [reply]
I am one of the contributors of the https://github.com/dcspy/dcspy. We implemented just enough features to be able to run it as a lambda function in AWS. So far, it seems to be doing okay for us (pulling 12 hours of data, one station at a time, all pseudo-binary messages, etc.). However, I'm sure there are several bugs and potential improvements. Anyway, I'm glad it's been somewhat useful to others as well.
